### PR TITLE
Refactor settings management

### PR DESCRIPTION
**Pull request by ciprianjichici**
# Refactor settings management

## The issue or feature being addressed

Align the management of settings with the new AIModel and APIEndpoints resources.
Have a single source of the truth for configuration between development and infrastructure.

## Details on the issue fix or feature implementation

N/A

## Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have included inline docs for my changes, where applicable
- [x]  I have successfully run a local build
- [ ]  I have provided the required update scripts, where applicable
- [ ]  I have updated relevant docs, where applicable

> [!NOTE]
> Instead of adding `X`'s inside the checkboxes you wish to check above, first submit the PR, then check the boxes in the rendered description.

### CODE REVIEW

File: src/dotnet/Common/Constants/Data/AppConfiguration.json [hunk 1]
```diff
@@ -0,0 +1,1494 @@
...
+        "secret": "",
+        "value": "#fff",
+        "content_type": "",
+        "first_version": "0.8.0"
+      },
+      {
+        "name": "FooterText",
+        "description": "Footer text.",
+        "secret": "",
+        "value": "FoundationaLLM (c) 2023",
```
Comment (joelhulen) [inline src/dotnet/Common/Constants/Data/AppConfiguration.json:1196]: We should set the year to 2024

File: src/dotnet/Common/Templates/appconfig.template.json [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+        {
+            "key": "FoundationaLLM:Branding:SecondaryButtonTextColor",
+            "value": "#fff",
+            "label": null,
+            "content_type": "",
+            "tags": {}
+        },
+        {
+            "key": "FoundationaLLM:Branding:FooterText",
+            "value": "FoundationaLLM (c) 2023",
```
Comment (joelhulen) [inline src/dotnet/Common/Templates/appconfig.template.json:663]: We should set the year to 2024

**Review by hatboyzero — APPROVED** [review]

File: src/dotnet/Common/Templates/appconfig.template.json [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+            "content_type": "application/json",
+            "tags": {}
+        },
+        {
+            "key": "FoundationaLLM:Events:Profiles:VectorizationWorker",
+            "value": "${env:FOUNDATIONALLM_VECTORIZATION_WORKER_EVENT_GRID_PROFILE}",
+            "label": null,
+            "content_type": "application/json",
+            "tags": {}
+        },
```
Comment (hatboyzero) [inline src/dotnet/Common/Templates/appconfig.template.json:821]: Need to resolve this trailing comma in the JSON dictionary.

**Review by hatboyzero — CHANGES_REQUESTED** [review]

File: src/dotnet/Common/Templates/appconfig.template.json [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+        {
+            "key": "FoundationaLLM:APIEndpoints:CoreAPI:APIUrl",
+            "value": "${env:SERVICE_CORE_API_ENDPOINT_URL}",
+            "label": null,
+            "content_type": "",
+            "tags": {}
+        },
+        {
+            "key": "FoundationaLLM:APIEndpoints:CoreAPI:AppInsightsConnectionString",
+            "value": "{\"uri\":\"${env:AZURE_KEY_VAULT_ENDPOINT}secrets/foundationallm-appinsights-connectionstring}\"}",
```
Comment (hatboyzero) [inline src/dotnet/Common/Templates/appconfig.template.json:138]: The trailing `}` in the keyvault URI before the closing quote `\"` is invalid and needs to be removed.  This is occurring in all keyvault URIs generated in this file.

**Review by hatboyzero — CHANGES_REQUESTED** [review]

**Review by hatboyzero — CHANGES_REQUESTED** [review]

File: deploy/quick-start/data/resource-provider/FoundationaLLM.Configuration/AzureEventGrid.template copy.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{
```
[the remark this answers is not in the document]
Comment (ciprianjichici) [reply deploy/quick-start/data/resource-provider/FoundationaLLM.Configuration/AzureEventGrid.template copy.json]: That is correct. Renamed the file.

File: src/dotnet/Common/Templates/azure-key-vault-secret-names.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    "FoundationaLLM:APIEndpoints:ManagementAPI:AppInsightsConnectionString": "foundationallm-appinsights-connectionstring",
+    "FoundationaLLM:APIEndpoints:VectorizationAPI:APIKey": "foundationallm-apiendpoints-vectorizationapi-apikey",
+    "FoundationaLLM:APIEndpoints:VectorizationAPI:AppInsightsConnectionString": "foundationallm-appinsights-connectionstring",
+    "FoundationaLLM:APIEndpoints:VectorizationWorker:APIKey": "foundationallm-apiendpoints-vectorizationworker-apikey",
+    "FoundationaLLM:APIEndpoints:VectorizationWorker:AppInsightsConnectionString": "foundationallm-appinsights-connectionstring",
+    "FoundationaLLM:APIEndpoints:GatewayAPI:APIKey": "foundationallm-apiendpoints-gatewayapi-apikey",
+    "FoundationaLLM:APIEndpoints:GatewayAPI:AppInsightsConnectionString": "foundationallm-appinsights-connectionstring",
+    "FoundationaLLM:APIEndpoints:GatewayAdapterAPI:APIKey": "foundationallm-apiendpoints-gatewayadapterapi-apikey",
+    "FoundationaLLM:APIEndpoints:GatewayAdapterAPI:AppInsightsConnectionString": "foundationallm-appinsights-connectionstring",
+    "FoundationaLLM:APIEndpoints:StateAPI:APIKey": "foundationallm-apinedpoints-stateapi-apikey",
```
Comment (hatboyzero) [inline src/dotnet/Common/Templates/azure-key-vault-secret-names.json:23]: The key name for State API in incorrectly spelled.  It should be `foundationallm-apiendpoints-stateapi-apikey`

**Review by hatboyzero — CHANGES_REQUESTED** [review]

File: src/dotnet/Common/Templates/appconfig.template.json [hunk 7]
```diff
@@ -0,0 +1,823 @@
...
+        {
+            "key": "FoundationaLLM:APIEndpoints:GatekeeperAPI:APIKey",
+            "value": "{\"uri\":\"${env:AZURE_KEY_VAULT_ENDPOINT}secrets/foundationallm-apiendpoints-gatekeeperapi-apikey}\"}",
+            "label": null,
+            "content_type": "application/vnd.microsoft.appconfig.keyvaultref+json;charset=utf-8",
+            "tags": {}
+        },
+        {
+            "key": "FoundationaLLM:APIEndpoints:GatekeeperAPI:AppInsightsConnectionString",
+            "value": "appinsights-connectionstring",
```
Comment (hatboyzero) [inline src/dotnet/Common/Templates/appconfig.template.json:243]: This value should be a secret like the rest of the service settings around AppInsights connection strings:

`{\"uri\":\"${env:AZURE_KEY_VAULT_ENDPOINT}secrets/foundationallm-apiendpoints-gatekeeperapi-apikey}\"}'

File: src/dotnet/Common/Templates/appconfig.template.json [hunk 8]
```diff
@@ -0,0 +1,823 @@
...
+            "label": null,
+            "content_type": "",
```
Comment (hatboyzero) [inline src/dotnet/Common/Templates/appconfig.template.json:245]: This value should specify a secret type like the rest of the service settings around AppInsights connection strings:

`application/vnd.microsoft.appconfig.keyvaultref+json;charset=utf-8`

**Review by hatboyzero — CHANGES_REQUESTED** [review]

**Comment by hatboyzero** [comment]
The following configuration settings are not present in the generated template:
- `FoundationaLLM:ResourceProviders:*:Storage:AccountName`
- `FoundationaLLM:ResourceProviders:*:Storage:AuthenticationType`

**Comment by hatboyzero** [comment]
DataSourceHub API is still looking for the following config setting:

`FoundationaLLM:APIs:DataSourceHubAPI:AppInsightsConnectionString`

**Comment by hatboyzero** [comment]
LangChainAPI is still looking for the following config setting:

`FoundationaLLM:APIs:LangChainAPI:AppInsightsConnectionString`

**Comment by hatboyzero** [comment]
PromptHub API is still looking for the following config setting:

`FoundationaLLM:APIs:PromptHubAPI:AppInsightsConnectionString`

**Comment by hatboyzero** [comment]
Services are crashing on invalid storage account names and invalid AppInsights connection strings.  I suspect that some of the settings still do not match up with what is expected.

**Review by hatboyzero — APPROVED** [review]

**Pull request merged**